packages/coding-style/src/Rector/ClassMethod/AnnotateThrowablesRector.php

```diff
@@ -0,0 +1,219 @@
+{
+    /**
+     * This is a comment.
+     *
+     * @param int $code
+     */
+    public function throwException(int $code)
+    {
+        throw new \RuntimeException('', $code);
+    }
+}
+PHP
+                    ,
+                    // code after
+                    <<<'PHP'
+class RootExceptionInMethodWithDocblock
+{
+    /**
+     * This is a comment.
+     *
+     * @param int $code
+     * @throws \RuntimeException
+     */
+    public function throwException(int $code)
+    {
+        throw new \RuntimeException('', $code);
+    }
+}
+PHP
+                ),
+            ]
+        );
+    }
+
+    /**
+     * @param Throw_ $node
+     */
+    public function refactor(Node $node): ?Node
+    {
+        if ($this->isThrowableAnnotated($node)) {
+            return null;
+        }
+
+        $this->annotateThrowable($node);
+
+        return $node;
+    }
+
+    private function isThrowableAnnotated(Throw_ $node): bool
+    {
+        $identifiedThrownThrowables = $this->identifyThrownThrowables($node);
+        $phpDocInfo = $this->getThrowingStmtDocblock($node);
+        $alreadyAnnotatedThrowTags = $phpDocInfo->getTagsByName('throws');
+
+        if (empty($alreadyAnnotatedThrowTags)) {
+            return false;
+        }
+
+        $alreadyAnnotatedThrowClassNames = array_map(static function($alreadyAnnotatedThrowTag) {
+            return $alreadyAnnotatedThrowTag->value->type->name;
+        }, $alreadyAnnotatedThrowTags);
+
+        $notAnnotatedThrowables = [];
+        foreach ($identifiedThrownThrowables as $identifiedThrownThrowable) {
+            foreach ($alreadyAnnotatedThrowClassNames as $alreadyAnnotatedThrowClassName) {
+                if ($alreadyAnnotatedThrowClassName === $identifiedThrownThrowable) {
+                    break;
+                }
+
+                if (
+                    ! Strings::contains($alreadyAnnotatedThrowClassName, '\\') &&
+                    Strings::contains($identifiedThrownThrowable, $alreadyAnnotatedThrowClassName) &&
+                    $this->isThrowableImported($node)
+                ) {
+                    break;
+                }
+            }
+
+            $notAnnotatedThrowables[] = $identifiedThrownThrowable;
+        }
+
+        return false === empty($notAnnotatedThrowables);
+    }
+
+    private function identifyThrownThrowables(Throw_ $node):array
+    {
+        switch (get_class($node->expr)) {
+            case New_::class:
+                return [$this->buildFQN($node)];
+            case StaticCall::class:
+                return $this->identifyThrownThrowablesInStaticCall($node);
+            default:
+                dumpf(null, $node);
+                throw new ShouldNotHappenException(sprintf('The \Throwable %s is not supported. Please, open an issue.', get_class($node->expr)));
+        }
+    }
+
+    private function identifyThrownThrowablesInStaticCall(Throw_ $node):array
+    {
+        ddf(null, 'Nodo' , $node);
+    }
```

Comment:
The full code is this.

If I find a `New_` node, then it is a throw like `throw new \Exception()`: non problem in this case.

But if the exception is thrown as `throw ExceptionFactory::createException()`, then I need to access the `createException()` method and read its docblock to see if the return type is annotated and, if not, read its return type hint.
The results of the Docblock `@return` annotation or, alternatively, of the return type hint have to then be used to annotate the method/function to which the analyzing `Throw_` node belongs to.

In case of static methods, I get a `StaticCall` node: from this I cannot find a path to access the node that pertains to `ExceptionFactory::createException()`: I can read the class name ´ExceptionFactory` and the method name `createException`, but cannot access their nodes.

Without being able to access their nodes I cannot know what is returned by `ExceptionFactory::createException()` and so I cannot annotate the `Throw_` node.

Reply:
In that case, I'd try `$this->getType($node)`

Reply:
@TomasVotruba , if you download the PR (did you see this https://github.blog/2020-02-12-supercharge-your-command-line-experience-github-cli-is-now-in-beta/ ?), you can better understand which is the problem.

If you run `vendor/bin/phpunit --filter AnnotateThrowables`, the `dumpf()` call will create an HTML file useful to navigate the full `Throw_ $node`: you will see it seems there is no path toward the `ExceptionFactory::createException()` node... or at least I didn't found it 😅

Reply:
For me, the PR is getting to huge to work effectively. We usually have 200-300 lines tests included.

I'd like to merge basic feature now to `master`, then resolve the edge cases that don't pass yet.

Could you move those non passing fixtures to `/FixtureTodo` diretory?

Reply:
I'll check the Github stuff, looks neat, thanks :+1:

Reply:
> For me, the PR is getting to huge to work effectively. We usually have 200-300 lines tests included.
> 
> I'd like to merge basic feature now to `master`, then resolve the edge cases that don't pass yet.
> 
> Could you move those non passing fixtures to `/FixtureTodo` diretory?

Ok, let me play for a bit more time: tomorrow I will send a working PR with the working rector.

Reply:
I'll be ready to merge :)